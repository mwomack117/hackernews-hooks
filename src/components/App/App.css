.page {
  margin: 20px;
  background-color: #f6f6ef;
}

.header {
  background-color: rgb(255, 102, 0);
  padding: 5px;
}

.nav-links {
  display: flex;
}

.nav-links li a {
  font-size: 16px;
  list-style: none;
  padding-right: 7px;
  text-decoration: none;
}

nav {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

nav h2 {
  padding-left: 10px;
}

input {
  font-size: 14px;
  width: 175px;
}

button {
  font-size: 14px;
}

form {
  text-align: center;
  margin-top: 5px;
}
.interactions {
  text-align: center;
}

.interactions > button {
  margin: 0 10px;
}

.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 18px;
  padding: 0 10px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.table-header > span,
.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.button-inline {
  border-width: 0;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 4px;
  font-size: inherit;
  cursor: pointer;
}

.button-reverse {
  display: flex;
  align-items: center;
  border: 1px rgb(255, 102, 0) solid;
  background: transparent;
  color: inherit;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 4px;
  font-size: inherit;
  cursor: pointer;
}

.button-reverse:focus,
.button-inline:focus {
  outline: none;
}

.button-active,
.button-active:hover {
  background-color: rgb(255, 102, 0);
  color: white;
  text-align: inherit;
  -webkit-font-smoothing: inherit;
  padding: 4px;
  font-size: inherit;
  cursor: pointer;
}
