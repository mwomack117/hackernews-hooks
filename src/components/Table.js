import React from "react";
import PropTypes from "prop-types";

import Button from "./Button";

const largeColumn = {
  width: "40%"
};

const midColumn = {
  width: "27%"
};

const smallColumn = {
  width: "11%"
};

const Table = ({ list, onDismiss }) => (
  <div className="table">
    <div className="table-header">
      <span style={largeColumn}>Article Title</span>
      <span style={midColumn}>Author</span>
      <span style={smallColumn}>Comments</span>
      <span style={smallColumn}>Points</span>
      <span style={smallColumn}></span>
    </div>
    {list.map(item => (
      <div key={item.objectID} className="table-row">
        <span style={largeColumn}>
          <a href={item.url}>{item.title}</a>
        </span>
        <span style={midColumn}>{item.author}</span>
        <span style={smallColumn}>{item.num_comments}</span>
        <span style={smallColumn}>{item.points}</span>
        <span style={smallColumn}>
          <Button
            onClick={() => onDismiss(item.objectID)}
            className="button-inline"
          >
            Dismiss
          </Button>
        </span>
      </div>
    ))}
  </div>
);

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired
};

Button.defaultProps = { className: "" };

export default Table;
