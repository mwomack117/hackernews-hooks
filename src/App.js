import React, { useState, useEffect } from "react";
import "./App.css";
import axios from "axios";
import PropTypes from "prop-types";

import Button from "./components/Button";
import Table from "./components/Table";
import Search from "./components/Search";

import {
  PATH_BASE,
  PARAM_HITS_PER_PAGE,
  PARAM_QUERY,
  PARAM_PAGE
} from "./constants";

const App = () => {
  const [results, setResults] = useState({});
  const [searchTerm, setSearchTerm] = useState(PARAM_QUERY);
  const [searchKey, setSearchKey] = useState("");
  const [error, setError] = useState(null);

  const needsToSearchTopStories = searchTerm => {
    return !results[searchTerm];
  };

  const setSearchTopStories = result => {
    const { hits, page } = result;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];

    const newResults = { ...results, [searchKey]: { hits: updatedHits, page } };
    setResults(newResults);
  };

  const fetchSearchTopStories = (searchTerm, page = 0) => {
    axios
      .get(PATH_BASE + searchTerm + PARAM_HITS_PER_PAGE + PARAM_PAGE + page)
      .then(response => {
        setSearchTopStories(response.data);
        console.log(response.data);
        console.log("api-hit");
      })
      .catch(error => {
        setError(error);
        console.log(error);
      });
  };

  useEffect(() => {
    setSearchKey(searchTerm);
    fetchSearchTopStories(searchTerm);
  }, []);

  const onSearchChange = event => {
    setSearchTerm(event.target.value);
  };

  const onSearchSubmit = event => {
    setSearchKey(searchTerm);

    if (needsToSearchTopStories(searchTerm)) {
      fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  };

  const onDismiss = id => {
    const isNotId = item => item.objectID !== id;
    const updatedHits = results[searchKey].hits.filter(isNotId);
    // same as ^^ // const updatedHits = result.hits.filter(item => item.objectID !== id);
    setResults({ ...results, [searchKey]: { hits: updatedHits } });
  };

  const page = (results && results[searchKey] && results[searchKey].page) || 0;
  const list = (results && results[searchKey] && results[searchKey].hits) || [];

  return (
    <div className="page">
      <p>Learn React</p>
      {console.log(results)}
      <div className="interactions">
        <Search
          value={searchTerm}
          onChange={onSearchChange}
          onSubmit={onSearchSubmit}
        >
          Search
        </Search>
      </div>
      {error ? (
        <div className="interactions">
          <p>
            Something went wrong{" "}
            <span role="img" aria-label="confused imoji">
              ðŸ¤”
            </span>
          </p>
        </div>
      ) : (
        <Table list={list} onDismiss={onDismiss} />
      )}
      <div className="interactions">
        <Button onClick={() => fetchSearchTopStories(searchKey, page + 1)}>
          More
        </Button>
        <p>Page: {page + 1}</p>
      </div>
    </div>
  );
};

Button.propTypes = {
  onClick: PropTypes.func.isRequired
};

Search.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired
};

export default App;
