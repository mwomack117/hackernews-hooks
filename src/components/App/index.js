import React, { useState, useEffect } from "react";
import "./App.css";
import axios from "axios";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import PropTypes from "prop-types";

import SignUpPage from "../SignUp";
import SignInPage from "../SignIn";
import PasswordForgetPage from "../PasswordForget";
import HomePage from "../Home";
import AccountPage from "../Account";

import * as ROUTES from "../../constants/routes";

import Button from "../Button";
import Table from "../Table";
import Search from "../Search";
import Header from "../Header";
import Loading from "../Loading";

import {
  PATH_BASE,
  PARAM_HITS_PER_PAGE,
  PARAM_QUERY,
  PARAM_PAGE
} from "../../constants/apiParams.js";

const App = () => {
  const [results, setResults] = useState({});
  const [searchTerm, setSearchTerm] = useState(PARAM_QUERY);
  const [searchKey, setSearchKey] = useState(PARAM_QUERY);
  const [error, setError] = useState(null);
  const [isLoading, setLoading] = useState(false);

  const needsToSearchTopStories = searchTerm => {
    return !results[searchTerm];
  };

  const setSearchTopStories = result => {
    const { hits, page } = result;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];

    const updatedHits = [...oldHits, ...hits];

    const newResults = { ...results, [searchKey]: { hits: updatedHits, page } };

    setResults(newResults);
    setLoading(false);
  };

  const fetchSearchTopStories = (searchTerm, page = 0) => {
    setLoading(true);

    axios
      .get(PATH_BASE + searchTerm + PARAM_HITS_PER_PAGE + PARAM_PAGE + page)
      .then(response => {
        setSearchTopStories(response.data);
        console.log("API call", response.data);
      })
      .catch(error => {
        setError(error);
        console.log(error);
      });
  };

  useEffect(() => {
    setSearchKey(searchTerm);

    if (needsToSearchTopStories(searchTerm)) {
      fetchSearchTopStories(searchTerm);
    }
  }, [searchKey]);

  const onSearchChange = event => {
    setSearchTerm(event.target.value);
  };

  const onSearchSubmit = event => {
    setSearchKey(searchTerm);

    event.preventDefault();
  };

  const onDismiss = id => {
    const isNotId = item => item.objectID !== id;
    const updatedHits = results[searchKey].hits.filter(isNotId);
    // same as ^^ // const updatedHits = results[searchKey].hits.filter(item => item.objectID !== id);
    setResults({ ...results, [searchKey]: { hits: updatedHits } });
  };

  const page = (results && results[searchKey] && results[searchKey].page) || 0;
  const list = (results && results[searchKey] && results[searchKey].hits) || [];

  return (
    <Router>
      <div className="page">
        <Header
          value={searchTerm}
          onChange={onSearchChange}
          onSubmit={onSearchSubmit}
        >
          Search
        </Header>

        <Switch>
          <Route exact path={"/"}>
            <Search
              value={searchTerm}
              onChange={onSearchChange}
              onSubmit={onSearchSubmit}
            >
              Search
            </Search>
            {error ? (
              <div className="interactions">
                <p>
                  Something went wrong{" "}
                  <span role="img" aria-label="confused imoji">
                    ðŸ¤”
                  </span>
                </p>
              </div>
            ) : (
              <Table list={list} onDismiss={onDismiss} />
            )}
            <div className="interactions">
              {isLoading ? (
                <Loading />
              ) : (
                <Button
                  onClick={() => fetchSearchTopStories(searchKey, page + 1)}
                >
                  More
                </Button>
              )}
            </div>
          </Route>
          <Route path={ROUTES.SIGN_UP} component={SignUpPage} />
          <Route path={ROUTES.SIGN_IN} component={SignInPage} />
          <Route path={ROUTES.PASSWORD_FORGET} component={PasswordForgetPage} />
          <Route path={ROUTES.HOME} component={HomePage} />
          <Route path={ROUTES.ACCOUNT} component={AccountPage} />
        </Switch>
      </div>
    </Router>
  );
};

Button.propTypes = {
  onClick: PropTypes.func.isRequired
};

Search.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired
};

export default App;
