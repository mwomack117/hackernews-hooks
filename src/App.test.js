import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App from "./App";
import Button from "./components/Button";
import Search from "./components/Search";
import Table from "./components/Table";
import { unmountComponentAtNode } from "react-dom";

Enzyme.configure({ adapter: new Adapter() });

describe("Button", () => {
  test("button element renders", () => {
    const wrapper = shallow(<Button />);

    expect(wrapper);
  });
});
