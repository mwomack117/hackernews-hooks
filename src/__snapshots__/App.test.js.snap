// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  type="button"
>
  Give me more
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  Search
   
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      Article Title
    </span>
    <span
      style={
        Object {
          "width": "27%",
        }
      }
    >
      Author
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      Comments
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      Points
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    />
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "27%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "27%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "11%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
